price-feeder/oracle/client/client.go

```diff
@@ -0,0 +1,274 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+	tmjsonclient "github.com/tendermint/tendermint/rpc/jsonrpc/client"
+	umeeapp "github.com/umee-network/umee/app"
+	umeeparams "github.com/umee-network/umee/app/params"
+)
+
+type (
+	CosmosChain struct {
+		ChainID             string
+		KeyringBackend      string
+		KeyringDir          string
+		KeyringPass         string
+		TMRPC               string
+		RPCTimeout          time.Duration
+		OracleAddr          sdk.AccAddress
+		OracleAddrString    string
+		ValidatorAddr       sdk.ValAddress
+		ValidatorAddrString string
+		Encoding            umeeparams.EncodingConfig
+		GasPrices           string
+		GasAdjustment       float64
+		GRPCEndpoint        string
+	}
+
+	passReader struct {
+		pass string
+		buf  *bytes.Buffer
+	}
+)
+
+func newPassReader(pass string) io.Reader {
+	return &passReader{
+		pass: pass,
+		buf:  new(bytes.Buffer),
+	}
+}
+
+func (r *passReader) Read(p []byte) (n int, err error) {
+	n, err = r.buf.Read(p)
+	if err == io.EOF || n == 0 {
+		r.buf.WriteString(r.pass + "\n")
+
+		n, err = r.buf.Read(p)
+	}
+
+	return n, err
+}
+
+type OracleClient struct {
+	CosmosChain       CosmosChain
+	KeyringPassphrase string
+	Ctx               client.Context
+	Factory           tx.Factory
+}
+
+// Pre-vote and vote are separated out mainly for readability
+// Prevote doesn't need the timeout functionality that vote needs,
+// Because of the block timing validation on the node side
+
+// Ref : https://github.com/terra-money/oracle-feeder/blob/baef2a4a02f57a2ffeaa207932b2e03d7fb0fb25/feeder/src/vote.ts#L230
+
+func (oc OracleClient) BroadcastPrevote(msgs ...sdk.Msg) error {
+	return tx.BroadcastTx(oc.Ctx, oc.Factory, msgs...)
+}
+
+// Need to try this continually, since err is either timing issue or whitelist error
```

Comment:
I don't understand the godoc here.

Reply:
Should probably ref https://github.com/terra-money/oracle-feeder/blob/baef2a4a02f57a2ffeaa207932b2e03d7fb0fb25/feeder/src/vote.ts#L230 or https://docs.terra.money/Reference/Terra-core/Module-specifications/spec-oracle.html#prevote-and-vote instead here to make more sense. The method has to perform re-tries because the voting has to occur in the voting period after the prevote, and sometimes will try this before next period begins